File: 백준/Silver/10451. 순열 사이클/순열 사이클.py
import sys
sys.setrecursionlimit(10**6)

def dfs(n):
    if marked[n]: return
    marked[n] = True
    for x in edge[n]:dfs(x)
    
    
t = int(input())

while t!=0:
    cnt=0
    edge = [[] for _ in range(1001)]
    marked = [False] * 1001
    n = int(input())
    x = list(map(int,input().split()))
    for i in range(1,n+1):
        edge[i].append(x[i-1])
        edge[x[i-1]].append(i)
    for i in range(1,n+1):
        if not marked[i]:
            dfs(i)
            cnt+=1
    print(cnt)
    t-=1

File: 백준/Silver/10451. 순열 사이클/순열 사이클.cs
using System;
using System.Text;
using System.Collections.Generic;

class Program
{
    static List<List<int>> edge = new List<List<int>>();
    static bool[] marked = new bool[1001];
    static int cnt = 0;
    
    static void dfs(int n) 
    {
        if(marked[n]) return;
        marked[n] = true;
        foreach (int x in edge[n]) dfs(x);
    }
    
    static void Main() {
        int n,t = int.Parse(Console.ReadLine());
        
        
        while(t != 0)
        {
            for(int i=0;i<1001;i++)
            {
                edge.Add(new List<int>());
            }
            
            n = int.Parse(Console.ReadLine());
            string[] str = Console.ReadLine().Split(' ');
            
            for(int i=0;i<n;i++)
            {
                edge[i+1].Add(int.Parse(str[i]));
                edge[int.Parse(str[i])].Add(i+1);
            }
            for(int i=1;i<=n;i++)
            {
                if(!marked[i])
                {
                    dfs(i);
                    cnt++;
                }
            }
            Console.WriteLine(cnt);
            
            edge.Clear();
            marked = new bool[1001];
            cnt=0;
            t--;
        }
    }
}